pkg/storage/ingest/pusher.go

```diff
@@ -103,15 +103,57 @@ func newPusherConsumer(pusher Pusher, kafkaCfg KafkaConfig, metrics *pusherConsu
 	}
 }
 
-func (c pusherConsumer) consume(ctx context.Context, records []record) error {
-	ctx, cancel := context.WithCancelCause(ctx)
-	defer cancel(cancellation.NewErrorf("done consuming records"))
+// Consume implements the recordConsumer interface.
+// It'll use a separate goroutine to unmarshal the next record while we push the current record to storage.
+func (c pusherConsumer) Consume(ctx context.Context, records []record) error {
+	recordsChannel := make(chan parsedRecord)
 
-	recC := make(chan parsedRecord)
+	// Create a cancellable context to let the unmarshalling goroutine know when to stop.
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
 
-	// Speed up consumption by unmarhsalling the next request while the previous one is being pushed.
-	go c.unmarshalRequests(ctx, records, recC)
-	return c.pushRequests(recC)
+	// Now, unmarshal the records into the channel.
+	go func(unmarshalCtx context.Context, records []record) {
+		defer close(recordsChannel)
+
+		for index, r := range records {
+			select {
+			case <-unmarshalCtx.Done():
+				// No more processing is needed, so we need to abort.
+				return
+			default:
+				parsed := parsedRecord{
+					ctx:          r.ctx,
+					tenantID:     r.tenantID,
+					WriteRequest: &mimirpb.WriteRequest{},
+					index:        index,
+				}
+
+				// We don't free the WriteRequest slices because they are being freed by a level below.
+				err := parsed.WriteRequest.Unmarshal(r.content)
+				if err != nil {
+					parsed.err = fmt.Errorf("parsing ingest consumer write request: %w", err)
+				}
+				recordsChannel <- parsed
+			}
+		}
+	}(ctx, records)
+
+	for r := range recordsChannel {
+		if r.err != nil {
+			level.Error(c.logger).Log("msg", "failed to parse write request; skipping", "err", r.err)
+			continue
+		}
+
+		// If we get an error at any point, we need to stop processing the records. They will be retried at some point.
+		err := c.pushToStorage(r.ctx, r.tenantID, r.WriteRequest)
+		if err != nil {
+			cancel() // Stop the unmarshalling goroutine.
```

Comment:
there's also a `defer cancel()`, do we need both?

Reply:
I've removed it in favour of manually calling `cancel` I think it has something to do with how defer works and our go verify leak where I get inconsistent results without calling cancel as we're done (and also the linter complains about it)